core/src/test/java/org/elasticsearch/threadpool/FixedThreadPoolTests.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.threadpool;
+
+import org.elasticsearch.common.settings.ClusterSettings;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.common.util.concurrent.EsRejectedExecutionException;
+
+import java.util.concurrent.CountDownLatch;
+
+import static org.hamcrest.CoreMatchers.equalTo;
+
+public class FixedThreadPoolTests extends ESThreadPoolTestCase {
+
+    public void testRejectedExecutionCounter() throws InterruptedException {
+        final String threadPoolName = randomThreadPool(ThreadPool.ThreadPoolType.FIXED);
+        // some of the fixed thread pool are bound by the number of
+        // cores so we can not exceed that
+        final int size = randomIntBetween(1, EsExecutors.boundedNumberOfProcessors(Settings.EMPTY));
+        final int queueSize = randomIntBetween(1, 16);
+        final long rejections = randomIntBetween(1, 16);
+
+        ThreadPool threadPool = null;
+        final Settings nodeSettings =
+            Settings.builder()
+                .put("node.name", "testRejectedExecutionCounter")
+                .put("threadpool." + threadPoolName + ".size", size)
+                .put("threadpool." + threadPoolName + ".queue_size", queueSize)
+                .build();
+        try {
+            threadPool = new ThreadPool(nodeSettings);
+
+            // these tasks will consume the thread pool causing further
+            // submissions to queue
+            final CountDownLatch latch = new CountDownLatch(size);
+            final CountDownLatch block = new CountDownLatch(1);
+            for (int i = 0; i < size; i++) {
+                threadPool.executor(threadPoolName).execute(() -> {
+                    try {
+                        latch.countDown();
+                        block.await();
+                    } catch (InterruptedException e) {
+                        fail(e.toString());
```

Comment:
I tried:

```diff
diff --git a/core/src/test/java/org/elasticsearch/threadpool/FixedThreadPoolTests.java b/core/src/test/java/org/elasticsearch/threadpool/FixedThreadPoolTests.java
index 85daebb..e599db0 100644
--- a/core/src/test/java/org/elasticsearch/threadpool/FixedThreadPoolTests.java
+++ b/core/src/test/java/org/elasticsearch/threadpool/FixedThreadPoolTests.java
@@ -103,4 +103,11 @@ public class FixedThreadPoolTests extends ESThreadPoolTestCase {
         }
     }
 
+    public void testFail() throws InterruptedException {
+        Thread t = new Thread(() -> fail(""));
+        t.start();
+        t.join();
+        assertTrue(true);
+    }
+
 }
```

```bash
$ 11:19:23 [jason:~/src/elastic/elasticsearch] rejected-execution-stats(+7/-0)+ 130 ± gradle :core:test -Dtests.class=org.elasticsearch.threadpool.FixedThreadPoolTests -Dtests.method=testFail
```

Gives:

```
.
.
.
:core:test
   [junit4] <JUnit4> says مرحبا! Master seed: 5EC46461968895CD
==> Test Info: seed=5EC46461968895CD; jvm=1; suite=1
Suite: org.elasticsearch.threadpool.FixedThreadPoolTests
  1> [2016-05-13 11:19:42,032][WARN ][org.elasticsearch.bootstrap] Unable to lock JVM Memory: error=78, reason=Function not implemented
  1> [2016-05-13 11:19:42,032][WARN ][org.elasticsearch.bootstrap] This can result in part of the JVM being swapped out.
  2> 五月 13, 2016 11:19:42 上午 com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
  2> WARNING: Uncaught exception in thread: Thread[Thread-1,5,TGRP-FixedThreadPoolTests]
  2> java.lang.AssertionError: 
  2>    at __randomizedtesting.SeedInfo.seed([5EC46461968895CD]:0)
  2>    at org.junit.Assert.fail(Assert.java:88)
  2>    at org.elasticsearch.threadpool.FixedThreadPoolTests.lambda$testFail$3(FixedThreadPoolTests.java:107)
  2>    at java.lang.Thread.run(Thread.java:745)
  2> REPRODUCE WITH: gradle :core:test -Dtests.seed=5EC46461968895CD -Dtests.class=org.elasticsearch.threadpool.FixedThreadPoolTests -Dtests.method="testFail" -Des.logger.level=WARN -Dtests.security.manager=true -Dtests.locale=zh -Dtests.timezone=America/Iqaluit
ERROR   0.14s | FixedThreadPoolTests.testFail <<< FAILURES!
   > Throwable #1: com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=14, name=Thread-1, state=RUNNABLE, group=TGRP-FixedThreadPoolTests]
   >    at __randomizedtesting.SeedInfo.seed([5EC46461968895CD:D367B5034637E5CC]:0)
   > Caused by: java.lang.AssertionError: 
   >    at __randomizedtesting.SeedInfo.seed([5EC46461968895CD]:0)
   >    at org.elasticsearch.threadpool.FixedThreadPoolTests.lambda$testFail$3(FixedThreadPoolTests.java:107)
   >    at java.lang.Thread.run(Thread.java:745)
  2> NOTE: test params are: codec=Asserting(Lucene60): {}, docValues:{}, maxPointsInLeafNode=1177, maxMBSortInHeap=6.435142190065553, sim=ClassicSimilarity, locale=zh, timezone=America/Iqaluit
  2> NOTE: Mac OS X 10.11.4 x86_64/Oracle Corporation 1.8.0_92 (64-bit)/cpus=8,threads=1,free=484700856,total=514850816
  2> NOTE: All tests run in this JVM: [FixedThreadPoolTests]
Completed [1/1] in 0.98s, 1 test, 1 error <<< FAILURES!

Tests with failures:
  - org.elasticsearch.threadpool.FixedThreadPoolTests.testFail

   [junit4] JVM J0:     0.31 ..     1.99 =     1.68s
   [junit4] Execution time total: 2.01 sec.
   [junit4] Tests summary: 1 suite, 1 test, 1 error
:core:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':core:test'.
> There were test failures: 1 suite, 1 test, 1 error [seed: 5EC46461968895CD]

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

Total time: 18.773 secs
```